### ISSUE TITLE
Crash in vkDestroyPipelineLayout

### ISSUE THREAD

**Issue by baldurk** [KhronosGroup/MoltenVK#705]
I ran into a crash inside vkDestroyPipelineLayout, which I found after some experimentation seemed to be related to destroying descriptor pools or maybe layouts before the pipeline layout.

In Vulkan there shouldn't be any ordering requirement on destruction, so it should be valid to destroy these in any order.

The only repro I can give is to load a capture of e.g. vkcube in RenderDoc and select the drawcall with the mesh viewer open - it will crash while fetching the mesh output data. I tried to modify vkcube itself to trigger this on shutdown but I couldn't get it to happen. I think it might be a use-after-free problem so it might require a more complex sample.

To repro the bug, you'll need to undo [the workaround I committed](https://github.com/baldurk/renderdoc/commit/4d99184626859444c7400c8fed609367cfba43e7).

```cpp
  m_pDriver->vkFreeDescriptorSets(dev, descpool, (uint32_t)descSets.size(), descSets.data());

  m_pDriver->vkDestroyDescriptorPool(dev, descpool, NULL);

  for(VkDescriptorSetLayout layout : setLayouts)
    m_pDriver->vkDestroyDescriptorSetLayout(dev, layout, NULL);

  // delete pipeline layout
  m_pDriver->vkDestroyPipelineLayout(dev, pipeLayout, NULL);

  // delete pipeline
  m_pDriver->vkDestroyPipeline(dev, pipe, NULL);

  // delete shader/shader module
  m_pDriver->vkDestroyShaderModule(dev, module, NULL);
```

<details>
<summary>Full backtrace</summary>

```
(lldb) bt
error: ReplayManager.o :: Class 'QThread' has a base class 'QObject' which does not have a complete definition.
error: ReplayManager.o :: Try compiling the source file with -fstandalone-debug.
* thread #12, name = 'QThread', stop reason = EXC_BAD_ACCESS (code=1, address=0x0)
  * frame #0: 0x000000012cbb0e82 libMoltenVK.dylib`std::__1::pair<std::__1::__hash_iterator<std::__1::__hash_node<std::__1::__hash_value_type<MVKDescriptorSetLayout*, MVKDeviceObjectPool<MVKDescriptorSet>*>, void*>*>, bool> std::__1::__hash_table<std::__1::__hash_value_type<MVKDescriptorSetLayout*, MVKDeviceObjectPool<MVKDescriptorSet>*>, std::__1::__unordered_map_hasher<MVKDescriptorSetLayout*, std::__1::__hash_value_type<MVKDescriptorSetLayout*, MVKDeviceObjectPool<MVKDescriptorSet>*>, std::__1::hash<MVKDescriptorSetLayout*>, true>, std::__1::__unordered_map_equal<MVKDescriptorSetLayout*, std::__1::__hash_value_type<MVKDescriptorSetLayout*, MVKDeviceObjectPool<MVKDescriptorSet>*>, std::__1::equal_to<MVKDescriptorSetLayout*>, true>, std::__1::allocator<std::__1::__hash_value_type<MVKDescriptorSetLayout*, MVKDeviceObjectPool<MVKDescriptorSet>*> > >::__emplace_unique_key_args<MVKDescriptorSetLayout*, std::__1::piecewise_construct_t const&, std::__1::tuple<MVKDescriptorSetLayout* const&>, std::__1::tuple<> >(this=0x000000012ecdcb98, __k=0x000070000cc7aa70, __args=0x000000012d02877b, __args=size=1, __args=size=0) at __hash_table:2020:16
    frame #1: 0x000000012cb9c531 libMoltenVK.dylib`std::__1::unordered_map<MVKDescriptorSetLayout*, MVKDeviceObjectPool<MVKDescriptorSet>*, std::__1::hash<MVKDescriptorSetLayout*>, std::__1::equal_to<MVKDescriptorSetLayout*>, std::__1::allocator<std::__1::pair<MVKDescriptorSetLayout* const, MVKDeviceObjectPool<MVKDescriptorSet>*> > >::operator[](this=0x000000012ecdcb98 size=1, __k=0x000070000cc7aa70) at unordered_map:1492:21
    frame #2: 0x000000012cb98fdb libMoltenVK.dylib`MVKDescriptorPool::removeDescriptorSetPool(this=0x000000012ecdcb40, mvkDescSetLayout=0x0000000109d91650) at MVKDescriptorSet.mm:1066:30
    frame #3: 0x000000012cb98e26 libMoltenVK.dylib`MVKDescriptorSetLayout::~MVKDescriptorSetLayout(this=0x0000000109d91650) at MVKDescriptorSet.mm:656:50
    frame #4: 0x000000012cb990e5 libMoltenVK.dylib`MVKDescriptorSetLayout::~MVKDescriptorSetLayout(this=0x0000000109d91650) at MVKDescriptorSet.mm:655:51
    frame #5: 0x000000012cc72b6a libMoltenVK.dylib`std::__1::enable_if<!(std::is_trivially_destructible<MVKDescriptorSetLayout>::value), void>::type mvk_vector_allocator_with_stack<MVKDescriptorSetLayout, 8>::destruct_all<MVKDescriptorSetLayout>(this=0x0000000109d91638) at MVKVectorAllocator.h:284:45
    frame #6: 0x000000012cc6d8d9 libMoltenVK.dylib`mvk_vector_allocator_with_stack<MVKDescriptorSetLayout, 8>::deallocate(this=0x0000000109d91638) at MVKVectorAllocator.h:540:5
    frame #7: 0x000000012cc6d967 libMoltenVK.dylib`mvk_vector_allocator_with_stack<MVKDescriptorSetLayout, 8>::~mvk_vector_allocator_with_stack(this=0x0000000109d91638) at MVKVectorAllocator.h:362:5
    frame #8: 0x000000012cc6d5d5 libMoltenVK.dylib`mvk_vector_allocator_with_stack<MVKDescriptorSetLayout, 8>::~mvk_vector_allocator_with_stack(this=0x0000000109d91638) at MVKVectorAllocator.h:361:3
    frame #9: 0x000000012cc74e11 libMoltenVK.dylib`MVKVectorImpl<MVKDescriptorSetLayout, mvk_vector_allocator_with_stack<MVKDescriptorSetLayout, 8> >::~MVKVectorImpl(this=0x0000000109d91628) at MVKVector.h:336:3
    frame #10: 0x000000012cc59345 libMoltenVK.dylib`MVKVectorImpl<MVKDescriptorSetLayout, mvk_vector_allocator_with_stack<MVKDescriptorSetLayout, 8> >::~MVKVectorImpl(this=0x0000000109d91628) at MVKVector.h:335:3
    frame #11: 0x000000012cc87ca0 libMoltenVK.dylib`MVKPipelineLayout::~MVKPipelineLayout(this=0x0000000109d91600) at MVKPipeline.h:45:7
    frame #12: 0x000000012cc62fe5 libMoltenVK.dylib`MVKPipelineLayout::~MVKPipelineLayout(this=0x0000000109d91600) at MVKPipeline.h:45:7
    frame #13: 0x000000012cc63009 libMoltenVK.dylib`MVKPipelineLayout::~MVKPipelineLayout(this=0x0000000109d91600) at MVKPipeline.h:45:7
    frame #14: 0x000000012cb9d16b libMoltenVK.dylib`MVKBaseObject::destroy(this=0x0000000109d91600) at MVKBaseObject.h:95:27
    frame #15: 0x000000012cbb71f8 libMoltenVK.dylib`MVKVulkanAPIObject::release(this=0x0000000109d91600) at MVKVulkanAPIObject.mm:32:49
    frame #16: 0x000000012cbb7245 libMoltenVK.dylib`MVKVulkanAPIObject::destroy(this=0x0000000109d91600) at MVKVulkanAPIObject.mm:36:2
    frame #17: 0x000000012cc4c424 libMoltenVK.dylib`MVKDevice::destroyPipelineLayout(this=0x000000010d887e00, mvkPLL=0x0000000109d91600, pAllocator=0x0000000000000000) at MVKDevice.mm:1948:10
    frame #18: 0x000000012cbd2e45 libMoltenVK.dylib`::vkDestroyPipelineLayout(device=0x000000010d887e18, pipelineLayout=0x0000000109d91600, pAllocator=0x0000000000000000) at vulkan.mm:1015:10
    frame #19: 0x000000010407e422 librenderdoc.dylib`WrappedVulkan::vkDestroyPipelineLayout(this=0x000000010a2ad000, device=0x000000010951ffd0, obj=0x000000010bb0ff10, pAllocator=0x0000000000000000) at vk_misc_funcs.cpp:111:1
    frame #20: 0x00000001039c35b1 librenderdoc.dylib`VulkanReplay::FetchVSOut(this=0x000000010a2ad130, eventId=11) at vk_postvs.cpp:2576:14
```
</details>

**Comment by billhollings** [comment]
I've encountered this too in other scenarios. I'm looking into it.

**Comment by billhollings** [comment]
@baldurk This is fixed in PR #714. Please retest and close if fixed.

**Comment by baldurk** [comment]
This is fixed now if I revert my workaround. Thanks!
